### PR TITLE
[CCP-168] Add tests

### PR DESCRIPTION
**Pull request by LomaxOnTheRun**
[JIRA ticket](https://commoncode.atlassian.net/browse/CCP-168)

This ticket is not finished but I'm moving off this project for a while so this is as far as I've gotten with it. Better than nothing. Comments for further work in JIRA ticket.

### CODE REVIEW

File: skills/__tests__/hears.test.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+const TEACHER_1_NAME = 'TEACHER_1_NAME';
+const TEACHER_2_ID = 'TEACHER2';
+const TEACHER_2_NAME = 'TEACHER_2_NAME';
+
+const mockBot = new (jest.fn())();
+mockBot.whisper = jest.fn();
+
+// These all get passed to asyncBot in extractMessageContents
+mockBot.api = jest.fn();
+mockBot.api.users = jest.fn();
```
Comment (tonytamps) [inline skills/__tests__/hears.test.js]: IMO the `asyncBot` mock should contain the `jest.fn()` implementations.
The tests can then just use `mockBot.api.users.mockImplementation() => { ... })` without needing to do the boilerplate.

Comment (tonytamps) [reply skills/__tests__/hears.test.js]: I'm sure you're aware already but be conscious that the mocks here are defined globally in the file and will apply to everything here with little ability to configure on a per-test basis.

File: skills/__tests__/hears.test.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  });
+
+  test('one skill and no teacher mentioned', () => {
+    const message = getMessage('Message with no teacher and one _skill_');
+    return expect(extractMessageContents(mockBot, message)).resolves.toMatchObject({
+      learnerObject: LEARNER_OBJECT,
+      teacherObjects: [],
+      skills: ['skill'],
+    });
+  });
```
Comment (tonytamps) [inline skills/__tests__/hears.test.js:59]: This is where I personally would have started following Arrange, Act, Assert for closely.
```javascript
  test('one skill and no teacher mentioned', async () => {
    const message = getMessage('Message with no teacher and one _skill_');

    const result = await extractMessageContents(mockBot, message)

    expect(result).toMatchObject({
      learnerObject: LEARNER_OBJECT,
      teacherObjects: [],
      skills: ['skill'],
    });
  });

**Review by tonytamps — APPROVED** [review]
Looks like a nice step forward 👌 great work.

File: skills/__tests__/hears.test.js (hunk 1 above)
Comment (LomaxOnTheRun) [reply skills/__tests__/hears.test.js]: I tried, but that led to my not being able to share constants across files (`TEST_USER_ID`, `TEST_USER_NAME`, etc.) which would lead to magic strings. Happy to hear about ways of fixing that, or if you think it's better to just have magic strings anyway.

And yes to the second point, but that didn't seems problematic at this stage as the mock functions fairly closely to how the function actually works. I would consider making it neater as and when it became problematic.

File: skills/__tests__/hears.test.js (hunk 2 above)
Comment (LomaxOnTheRun) [reply skills/__tests__/hears.test.js:59]: Yep, totally on board with this, gonna change them up to set a good precedence.

File: skills/__tests__/hears.test.js (hunk 1 above)
Comment (tonytamps) [reply skills/__tests__/hears.test.js]: if your `asyncBot.js` mock looks like this
```javascript
module.exports = {
  api: {
    users: { 
      info: jest.fn() 
    }
  }
};
```

Then each tests that uses new functions can continue to expand the representation of the mock as required.

All of the above lines go away and you can do this inside each tests that uses that function.

```javascript
mockBot.api.users.info.mockResolvedValue({ user: { name: TEST_USER_NAME } })
```

Since this function is presumably tested by you, or the package maintainer, proving that "when passes X returns Y" is outside the scope of these tests here.

Comment (LomaxOnTheRun) [reply skills/__tests__/hears.test.js]: I don't think that works because it's not the `mockBot` that gets directly used in the `extractMessageContents` function:

```const asyncBot = genAsyncBot(bot);```

The reason I have it set against my `mockBot` in the test is because my mocked `genAsyncBot` just passes back the bot that's given to it.

I _could_ sneak the return value in somewhere in the bot params, but that feels real dirty.

Comment (LomaxOnTheRun) [reply skills/__tests__/hears.test.js]: Tried a thing anyway, not sure how I feel about it though so happy to shoot in a different direction if you know of a better one.

File: lib/__mocks__/asyncBot.js [hunk 3]
```diff
@@ -1,10 +1,9 @@
...
     users: {
-      info: bot.asyncBotReturn || jest.fn(),
+      info: jest.fn(),
     },
   },
-});
+};
 
-module.exports = genAsyncBot;
+module.exports = jest.fn(() => mockAsyncBot);
```
Comment (tonytamps) [inline lib/__mocks__/asyncBot.js:9]: You _could_ take the `bot` param and do something with it, but I didn't see any need.
Easy change if someone wanted to do that later.

File: skills/__fixtures__/teachers.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  [teacher1.id]: teacher1,
+  [teacher2.id]: teacher2,
+};
+
+module.exports = {
+  teacher1,
+  teacher2,
+  teachersArr,
+  teachersObj,
+};
```
Comment (tonytamps) [inline skills/__fixtures__/teachers.js:16]: Same 2 fixtures arranged in a few ways, makes using/asserting with them in the tests easier. 
Also relates some of the "magic strings" together that were only related by naming convention.
Gives a nicer idea of what the object should look like, easier to spot issues if anything fails 

> Wait a minute... Teacher object doesn't have a `.whatever` prop!

File: skills/__tests__/hears.test.js [hunk 5]
```diff
@@ -77,46 +72,50 @@ describe('Test extractMessageContent', () => {
   });
 
   test('no skill and one teacher mentioned', async () => {
-    const message = getMessage(`Message with no skill and <@${TEACHER_1_ID}>`);
+    const message = getMessage(`Message with no skill and <@${teacher1.id}>`);
+    asyncBot.api.users.info.mockResolvedValueOnce({ user: { name: teacher1.name } });
+    asyncBot.api.users.info.mockResolvedValueOnce({ user: { name: TEST_USER_NAME } });
```
Comment (tonytamps) [inline skills/__tests__/hears.test.js:77]: Each tests performs it's own mocks because it makes it easier to mock fail cases / edge cases.
In this file it was harder than I'd like because of 2 things:

1. `extractMessageContents` is probably doing too much
2. `asyncBot` is actually a tricky mock to get right

File: lib/__mocks__/asyncBot.js (hunk 3 above)
Comment (tonytamps) [reply lib/__mocks__/asyncBot.js:9]: Also really important here, the reason this actually works as a mock is because the implementation is _passing by reference_ not creating a new object every time.

This would *not* work:
```javascript
module.exports = jest.fn(() => ({
  api: {
    users: {
      info: bot.asyncBotReturn || jest.fn(),
      info: jest.fn(),
    },
  },
});
}));
```

Comment (tonytamps) [reply lib/__mocks__/asyncBot.js:9]: It's all the same code but when the genAsyncBot function is executed in the test, the mocks are applied to it.
The same func is then run in the implementation and a new instance is returned not having any of the mocks.

Pass by ref allows us to mock the behaviour and have the config persist on subsequent calls. Make sense?

**Review by tonytamps — COMMENTED** [review]
@LomaxOnTheRun I found some time, take a look at this :)

**Review by tonytamps — APPROVED** [review]
Very nice, lets merge it and see what happens :D
